### ISSUE TITLE
[New-Resource] SPSecurityTokenServiceConfig

### ISSUE THREAD

**Issue by Yvand** [PowerShell/SharePointDsc#673]
Hello, I want to configure apps in a HTTP site, so I have to set property AllowOAuthOverHttp to true:
```powershell
$serviceConfig = Get-SPSecurityTokenServiceConfig
$serviceConfig.AllowOAuthOverHttp = $true
$serviceConfig.Update()
```
I didn't find any SharePointDsc resource to configure object SPSecurityTokenServiceConfig.
Then, how should I set this property?
Thanks for your input,
Yvan

**Comment by BrianFarnhill** [comment]
We don't have a resource for managing the STS config, but that's not to say we shouldn't. Let put this on the backlog and see if we can get someone to code something up.

**Comment by Yvand** [comment]
Ok, just a remark on this: it's not a resource that we can add or remove in the farm, it's just an object that can be configured (or left to its default configuration).
So it means that property "Ensure" Present/Missing is completely irrelevant here.
Is it still relevant to create a resource to configure this?

So should the name of the resource bo explicitely "Configure-SPSecurityTokenServiceConfig" instead of just "SPSecurityTokenServiceConfig"?

**Comment by BrianFarnhill** [comment]
@Yvand It actually is a DSC resource. Don't think of resources as something that I can add multiple instances of (like web apps or sites), it can be a generic farm-wide setting (like we have for SPDiagnosticLogging or SPOutgoingEmail). So, in this case, we can write it as a single instance resource that checks if the settings are correct, and then fixes them if they are not. It's not a complex resource at all, but definitely, suits the model.

**Comment by Yvand** [comment]
Ok, indeed, thanks for the explanation.
In the meantime, do I have a simple way to configure the STS as in my initial message?

**Comment by andikrueger** [comment]
You could use the [script resource](https://github.com/PowerShell/SharePointDsc/wiki/Using-the-Script-resource-in-configurations-that-use-SharePointDsc):
For example:
```
Script STS {
        GetScript = {
            return @{}
        }
        SetScript = {
            Invoke-SPDscCommand -ScriptBlock {
                $serviceConfig = Get-SPSecurityTokenServiceConfig
                $serviceConfig.AllowOAuthOverHttp = $true
                $serviceConfig.Update()
            }
        }
        TestScript = {
            return $false
        }
    }
```

**Comment by BrianFarnhill** [comment]
@andikrueger is right, but keep in mind this will apply the setting every time you run your config, regardless of the current state.

**Comment by Yvand** [comment]
Perfect, thank you for the sample and the piece of advice!

**Comment by Yvand** [comment]
@andikrueger your sample is great, but I'm trying to pass some parameters to the script and I'm not successful so far, despite my research.
This doesn't work:
```powershell
Script ConfigureSTSAndMultipleZones
{
	GetScript = { return @{}	}
	SetScript = {
		$waValue = $using:SPTrustedSitesName
		$argumentList = @()
		$argumentList += ("-wa", "`"$waValue`"")
		Invoke-SPDscCommand -Credential $SPSetupCredsQualified -Arguments @argumentList -ScriptBlock {
			 param(
				[string]$wa
			)
			
			$webAppUrl = "http://$wa/"
			(Get-SPWebApplication $webAppUrl).Url
		}
	}
	TestScript = { return $false }
	PsDscRunAsCredential = $SPSetupCredsQualified
}
```
I get the error "A positional parameter cannot be found that accepts argument '"SPSites"'. "
"SPSites" is the actual value of $using:SPTrustedSitesName

Do you know by any chance what I'm doing wrong?

**Comment by andikrueger** [comment]
I think there is an error in your argument list. You try to create sort of an array, but not a hashtable to pass the values to your script block. Could you try this:

```
        $argumentList = @{
            wa = "$($waValue)"
        }
```

**Comment by Yvand** [comment]
I found the solution, I'm happy to share it because it was not simple.
Here is how to pass parameters:
```powershell
Script ConfigureSTSAndMultipleZones
{
    GetScript = { return @{} }
    SetScript = {
        $argumentList = @(@{ "webAppUrl"             = "http://$using:SPTrustedSitesName"; 
                             "AppDomainFQDN"         = "$using:AppDomainFQDN"; 
                             "AppDomainIntranetFQDN" = "$using:AppDomainIntranetFQDN" })
        Invoke-SPDscCommand -Credential $DomainAdminCredsQualified -Arguments @argumentList -ScriptBlock {
            $params = $args[0]
            
            $webAppUrl = $params.webAppUrl
            $appDomainDefaultZone = $params.AppDomainFQDN
            $appDomainIntranetZone = $params.AppDomainIntranetFQDN
        }
    }
    TestScript = { return $false }
    PsDscRunAsCredential = $SPSetupCredsQualified
    DependsOn = "[SPSite]AppCatalog"
}
```

**Comment by NikCharlebois** [comment]
@ykuijs assigned that one to me. I will work on a new resource for an upcoming release

**Comment by NikCharlebois** [comment]
New resource has been created and is in my PR. Will be included in the next release on February 7

**Comment by ykuijs** [comment]
Closed via PR #729
